### PR TITLE
Add template classes for ROS2 action clients and servers

### PR DESCRIPTION
**Pull request by mjeronimo**
## Description of contribution in a few bullet points

* This PR adds a couple template classes, similar to the existing service_client.hpp, to allow for easy use of the ROS2 action clients and servers
* Also includes a unit test for these classes

## Future work that may be required in bullet points

* The next step will be to port task servers and clients to the ROS2 actions

### CODE REVIEW

**Comment by crdelsey** [comment]
Also, this is failing CI.

File: nav2_util/include/nav2_util/simple_action_client.hpp [hunk 1]
```diff
@@ -0,0 +1,178 @@
...
+
+  ActionStatus invoke(
+    const typename ActionT::Goal & goal,
+    typename rclcpp_action::ClientGoalHandle<ActionT>::WrappedResult & result,
+    typename rclcpp_action::ClientGoalHandle<ActionT>::FeedbackCallback callback = nullptr)
+  {
+    // Send the goal
+
+    auto future_goal_handle = action_client_->async_send_goal(goal, callback);
+
```
Comment (mhpanah) [inline nav2_util/include/nav2_util/simple_action_client.hpp]: Here, we should wait for the action return with some timeout.

File: nav2_util/include/nav2_util/simple_action_client.hpp [hunk 2]
```diff
@@ -0,0 +1,178 @@
...
+    if (!goal_handle) {
+      throw std::runtime_error("Goal was rejected by the server");
+    }
+
+    // Wait for the result
+
+    auto future_result = action_client_->async_get_result(goal_handle);
+
+    if (rclcpp::spin_until_future_complete(node_, future_result) !=
```
Comment (mhpanah) [inline nav2_util/include/nav2_util/simple_action_client.hpp]: We should wait with an option for timeout.

File: nav2_util/include/nav2_util/simple_action_server.hpp [hunk 3]
```diff
@@ -0,0 +1,100 @@
+// Copyright (c) 2018 Intel Corporation
```
Comment (mhpanah) [inline nav2_util/include/nav2_util/simple_action_server.hpp]: 2019

File: nav2_util/test/test_actions.cpp [hunk 4]
```diff
@@ -0,0 +1,331 @@
+// Copyright (c) 2018 Intel Corporation
```
Comment (mhpanah) [inline nav2_util/test/test_actions.cpp]: 2019

File: nav2_util/test/test_actions.cpp [hunk 5]
```diff
@@ -0,0 +1,331 @@
...
+  auto goal = Fibonacci::Goal();
+  goal.order = 20;
+
+  int sum = 0;
+
+  node_->action_client_->send_goal(goal);
+
+  bool cancel_received = false;
+
+  for (bool done = false; !done; ) {
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/test/test_actions.cpp]: I've removed the action client template, which turned out to not provide very much value over using the action client interface directly. I also found out that there will be changes to the action client class in ROS2 to make it simpler. So, the code reference has been removed along with the client template. .

File: nav2_util/include/nav2_util/simple_action_client.hpp [hunk 6]
```diff
@@ -0,0 +1,178 @@
+// Copyright (c) 2018 Intel Corporation
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_client.hpp]: I've removed the simple_action_client template since this turned out to not be very useful over using the action client interface directly. So, this file has been removed.

File: nav2_util/include/nav2_util/simple_action_server.hpp (hunk 3 above)
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_server.hpp]: Fixed.

File: nav2_util/test/test_actions.cpp (hunk 4 above)
Comment (mjeronimo) [reply nav2_util/test/test_actions.cpp]: Fixed.

File: nav2_util/include/nav2_util/simple_action_client.hpp [hunk 7]
```diff
@@ -0,0 +1,178 @@
...
+  {
+    return action_client_->wait_for_action_server(timeout);
+  }
+
+  ActionStatus invoke(
+    const typename ActionT::Goal & goal,
+    typename rclcpp_action::ClientGoalHandle<ActionT>::WrappedResult & result,
+    typename rclcpp_action::ClientGoalHandle<ActionT>::FeedbackCallback callback = nullptr)
+  {
+    // Send the goal
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_client.hpp]: I've deleted the action_client template (see other comments; turned out to not be very useful).

File: nav2_util/include/nav2_util/simple_action_client.hpp [hunk 8]
```diff
@@ -0,0 +1,178 @@
...
+
+    if (!goal_handle_) {
+      throw std::runtime_error("Goal was rejected by the server");
+    }
+  }
+
+  bool cancel()
+  {
+    auto future_cancel = action_client_->async_cancel_goal(goal_handle_);
+    return rclcpp::spin_until_future_complete(node_, future_cancel) ==
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_client.hpp]: I've deleted the action_client template (see other comments; turned out to not be very useful).

File: nav2_util/include/nav2_util/simple_action_client.hpp (hunk 2 above)
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_client.hpp]: I've deleted the action_client template (see other comments; turned out to not be very useful).

File: nav2_util/include/nav2_util/simple_action_client.hpp [hunk 9]
```diff
@@ -0,0 +1,178 @@
...
+    }
+  }
+
+  void send_goal(
+    const typename ActionT::Goal & goal,
+    typename rclcpp_action::ClientGoalHandle<ActionT>::FeedbackCallback callback = nullptr)
+  {
+    auto future_goal_handle = action_client_->async_send_goal(goal, callback);
+
+    if (rclcpp::spin_until_future_complete(node_, future_goal_handle) !=
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_client.hpp]: I've deleted the action_client template (see other comments; turned out to not be very useful).

File: nav2_util/test/test_actions.cpp [hunk 10]
```diff
@@ -0,0 +1,331 @@
...
+      if (current_goal_handle->is_canceling()) {
+        result->sequence = sequence;
+        current_goal_handle->set_canceled(result);
+        return;
+      }
+
+      // Check if we've gotten an new goal, pre-empting the current one
+      if (action_server_->update_requested()) {
+        current_goal_handle = action_server_->get_updated_goal_handle();
+        goto preempted;
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/test/test_actions.cpp]: Good point about the loop rate. That could be hoisted above the preempted: label. I'll make that change.

Regarding the suggested change: It's not just the goal that has to be updated, but in the general case, one must also clear the feedback and potentially the result (it could be possible that a result was being accumulated). So, there would need to be code to handle this as well; basically a duplication of the code that is already present above. So, I think that it's less error prone and a better template the way it is.

For example, this:

```C++
  void execute(const std::shared_ptr<GoalHandle> goal_handle)
  {
    // The goal may be pre-empted, so keep a pointer to the current goal
    std::shared_ptr<GoalHandle> current_goal_handle = goal_handle;

    rclcpp::Rate loop_rate(10);

preempted:
    // Initialize the goal, feedback, and result
    auto goal = current_goal_handle->get_goal();
    auto feedback = std::make_shared<Fibonacci::Feedback>();
    auto result = std::make_shared<Fibonacci::Result>();

    // Fibonacci-specific initialization
    auto & sequence = feedback->sequence;
    sequence.push_back(0);
    sequence.push_back(1);

    for (int i = 1; (i < goal->order) && rclcpp::ok(); ++i) {
      // Check if this action has been canceled
      ...

      // Check if we've gotten an new goal, pre-empting the current one
      if (action_server_->update_requested()) {
        current_goal_handle = action_server_->get_updated_goal_handle();
        goto preempted;
      }

      // Update the sequence
      sequence.push_back(sequence[i] + sequence[i - 1]);

      // Publish feedback
      current_goal_handle->publish_feedback(feedback);
      loop_rate.sleep();
    }
    ...
  }
```

Instead of this:

```C++
  void execute(const std::shared_ptr<GoalHandle> goal_handle)
  {
    // The goal may be pre-empted, so keep a pointer to the current goal
    std::shared_ptr<GoalHandle> current_goal_handle = goal_handle;

    rclcpp::Rate loop_rate(10);

    // Initialize the goal, feedback, and result
    auto goal = current_goal_handle->get_goal();
    auto feedback = std::make_shared<Fibonacci::Feedback>();
    auto result = std::make_shared<Fibonacci::Result>();

    // Fibonacci-specific initialization
    auto & sequence = feedback->sequence;
    sequence.push_back(0);
    sequence.push_back(1);

    for (int i = 1; (i < goal->order) && rclcpp::ok(); ++i) {
      // Check if this action has been canceled
      ...

      // Check if we've gotten an new goal, pre-empting the current one
      if (action_server_->update_requested()) {
        current_goal_handle = action_server_->get_updated_goal_handle();
        goal = current_goal_handle->get_goal();
        feedback = std::make_shared<Fibonacci::Feedback>();
        result = std::make_shared<Fibonacci::Result>(); 

        sequence = feedback->sequence;
        sequence.push_back(0);
        sequence.push_back(1);
      } else {
          // Update the sequence
          sequence.push_back(sequence[i] + sequence[i - 1]);

          // Publish feedback
          current_goal_handle->publish_feedback(feedback);
      }

      loop_rate.sleep();
    }
    ...
  }
```

[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/test/test_actions.cpp]: By the way, in some cases, where there isn't any state in the feedback and result, I agree with the change you propose. DWB, for example is like this. So, I think which way is cleaner will depend on the circumstances.

File: nav2_util/include/nav2_util/lifecycle_utils.hpp [hunk 11]
```diff
@@ -1,4 +1,4 @@
-// Copyright (c) 2018 Intel Corporation
+// Copyright (c) 2019 Intel Corporation
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/lifecycle_utils.hpp]: It was a new header, created in 2019, and I realized I copied the header from a file that had 2018. Good thing to clarify though.

File: nav2_util/include/nav2_util/simple_action_server.hpp [hunk 12]
```diff
@@ -0,0 +1,100 @@
...
+    auto handle_cancel = [](std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>>)
+      {
+        return rclcpp_action::CancelResponse::ACCEPT;
+      };
+
+    auto handle_accepted = [this](std::shared_ptr<rclcpp_action::ServerGoalHandle<ActionT>> handle)
+      {
+        // If we're currently working on a task, set a flag so that the
+        // action server can grab the pre-empting request in its loop
+        if (received_handle_ != nullptr && received_handle_->is_active()) {
```
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_server.hpp]: The server_goal_handle class itself serializes access with its own mutex: https://github.com/ros2/rclcpp/blob/master/rclcpp_action/src/server_goal_handle.cpp, which is a better place for it, IMO, so that clients of the class don't have to.

[the remark this answers is not in the document]
Comment (crdelsey) [reply nav2_util/include/nav2_util/simple_action_server.hpp]: It only serializes access to the server goal handle members. You've added additional members in this class that also need to be synced.

Comment (crdelsey) [inline nav2_util/include/nav2_util/simple_action_server.hpp]: It occurred to me last night that there might be another race condition here. The received_handle could change to inactive between your check at the top of this function and when the thread checks that an update is requested. In that case, a new thread would never be executed. This wouldn't be solved by the lock I suggested.

File: nav2_util/include/nav2_util/simple_action_server.hpp [hunk 13]
```diff
@@ -0,0 +1,100 @@
...
+          update_requested_ = true;
+          received_handle_ = handle;
+        } else {
+          // Otherwise, safe to start a new task
+          received_handle_ = handle;
+          std::thread{execute_callback_, handle}.detach();
```
Comment (crdelsey) [inline nav2_util/include/nav2_util/simple_action_server.hpp]: How do these threads get cleaned up in the end? I thought you had to `join` them to get them to finally go away.

Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_server.hpp]: The thread exits on its own after completing the execute callback. No join necessary for detacted threads (one must join _or_ detach).

File: nav2_util/include/nav2_util/simple_action_server.hpp (hunk 12 above)
[the remark this answers is not in the document]
Comment (mjeronimo) [reply nav2_util/include/nav2_util/simple_action_server.hpp]: Ok, true. I've added a mutex to serialize access.

**Comment by crdelsey** [comment]
I'm approving this as the PR because the race condition is going to possibly be addressed in a seperate PR
